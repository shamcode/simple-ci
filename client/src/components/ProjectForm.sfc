<template>
    <div class="project-create-form">
        <form :onsubmit={{this$._submitForm}} :ref="formNode">
            <div class="group">
                <input name="name" type="text" required value={{name}}/>
                <span class="highlight"></span>
                <span class="bar"></span>
                <label>Name</label>
            </div>
            <div class="group">
                <input name="cwd" type="text" required value={{cwd}}/>
                <span class="highlight"></span>
                <span class="bar"></span>
                <label>Current Work Directory</label>
            </div>
            <div class="group">
                <button type="submit" :disabled={{dataSaving}} >
                    <span>{{saveButtonText}}</span>
                </button>
            </div>
        </form>
    </div>
</template>

<script>

    function extractValue( value ) {
        return value.trim();
    }

    function validateValue( value ) {
        return value.length > 0;
    }


    function ProjectForm( options ) {
        const onSubmit = $();

        const state = options( {
            [ onSubmit ]() {},
            [ $.name ]: '',
            [ $.cwd ]: '',
            [ $.saveButtonText ]: '',
            [ $.dataSaving ]: false
        } );

        this$._submitForm = e => {
            e.preventDefault();
            const formData = new FormData( this.formNode );
            const data = {};
            let dataValid = true;
            [
                'name',
                'cwd'
            ].forEach( ( key ) => {
                const value = extractValue( formData.get( key ) );
                dataValid = dataValid && validateValue( value );
                if ( dataValid ) {
                    data[ key ] = value;
                }
            } );
            if ( dataValid ) {
                state[ onSubmit ]( data );
            }
        };
    }

    export default Component( Template, ProjectForm );
</script>
