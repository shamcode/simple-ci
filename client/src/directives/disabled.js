export default class {
    constructor() {
        this.node = null;
    }

    bind( node ) {
        this.node = node;
    }

    unbind() {
        this.node = null;
    }

    update( value ) {
        this.node.disabled = value;
    }
}
