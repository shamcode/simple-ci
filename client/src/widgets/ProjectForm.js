import { options } from 'sham-ui';
import ProjectFormTemplate from './ProjectForm.sht';

function extractValue( value ) {
    return value.trim();
}

function validateValue( value ) {
    return value.length > 0;
}

export default class ProjectForm extends ProjectFormTemplate {
    @options name = '';
    @options cwd = '';
    @options saveButtonText = '';
    @options dataSaving = false;
    @options onSubmit() {}

    _submitForm( e ) {
        e.preventDefault();
        const formData = new FormData( this.formNode );
        const data = {};
        let dataValid = true;
        [
            'name',
            'cwd'
        ].forEach( ( key ) => {
            const value = extractValue( formData.get( key ) );
            dataValid = dataValid && validateValue( value );
            if ( dataValid ) {
                data[ key ] = value;
            }
        } );
        if ( dataValid ) {
            this.options.onSubmit( data );
        }
    }
}
