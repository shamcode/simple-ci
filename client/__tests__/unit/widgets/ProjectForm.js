import ProjectForm from '../../../src/widgets/ProjectForm';
import directives from 'sham-ui-directives';
import renderer from 'sham-ui-test-helpers';

it( 'renders correctly', () => {
    const meta = renderer( ProjectForm, {
        directives
    } );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'submit form', () => {
    const onSubmitMock = jest.fn();
    const { widget } = renderer( ProjectForm, {
        saveButtonText: 'Create',
        directives,
        onSubmit: onSubmitMock
    } );

    const formData = {
        name: 'Test name',
        cwd: 'test cwd'
    };
    widget.querySelector( '[name="name"]' ).value = formData.name;
    widget.querySelector( '[name="cwd"]' ).value = formData.cwd;
    widget.querySelector( '[type="submit"]' ).click();

    expect( onSubmitMock.mock.calls ).toHaveLength( 1 );
    expect( onSubmitMock.mock.calls[ 0 ] ).toHaveLength( 1 );
    const data = onSubmitMock.mock.calls[ 0 ][ 0 ];
    expect( Object.keys( data ) ).toEqual( [ 'name', 'cwd' ] );
    expect( data.name ).toBe( formData.name );
    expect( data.cwd ).toBe( formData.cwd );
} );

it( 'form validation (name)', () => {
    const onSubmitMock = jest.fn();
    const { widget } = renderer( ProjectForm, {
        saveButtonText: 'Create',
        directives,
        onSubmit: onSubmitMock
    } );

    widget.querySelector( '[type="submit"]' ).click();
    expect( onSubmitMock.mock.calls ).toHaveLength( 0 );

    widget.querySelector( '[name="cwd"]' ).value = 'test';
    widget.querySelector( '[type="submit"]' ).click();
    expect( onSubmitMock.mock.calls ).toHaveLength( 0 );
} );

it( 'form validation (cwd)', () => {
    const onSubmitMock = jest.fn();
    const { widget } = renderer( ProjectForm, {
        saveButtonText: 'Create',
        directives,
        onSubmit: onSubmitMock
    } );

    widget.querySelector( '[type="submit"]' ).click();
    expect( onSubmitMock.mock.calls ).toHaveLength( 0 );

    widget.querySelector( '[name="name"]' ).value = 'test';
    widget.querySelector( '[type="submit"]' ).click();
    expect( onSubmitMock.mock.calls ).toHaveLength( 0 );
} );
