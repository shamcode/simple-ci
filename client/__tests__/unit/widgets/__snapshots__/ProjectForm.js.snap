// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Object {
  "ID": "widget",
  "constructorOptions": Object {
    "directives": Object {
      "EventListener": [Function],
      "disabled": [Function],
      "onclick": [Function],
      "oninput": [Function],
      "onload": [Function],
      "onresize": [Function],
      "onscroll": [Function],
      "onselect": [Function],
      "onsubmit": [Function],
    },
  },
  "html": "
<div class=\\"project-create-form\\">
  <form>
    <div class=\\"group\\">
      <input name=\\"name\\" type=\\"text\\" required=\\"\\"><span class=\\"highlight\\"></span><span class=\\"bar\\"></span>
      <label>Name</label>
    </div>
    <div class=\\"group\\">
      <input name=\\"cwd\\" type=\\"text\\" required=\\"\\"><span class=\\"highlight\\"></span><span class=\\"bar\\"></span>
      <label>Current Work Directory</label>
    </div>
    <div class=\\"group\\">
      <button type=\\"submit\\"><span></span></button>
    </div>
  </form>
</div>
",
  "nodes": Array [
    Object {
      "ID": undefined,
      "constructorOptions": null,
      "html": null,
      "nodes": null,
      "options": null,
    },
  ],
  "options": Object {
    "actionSequence": Array [
      "render",
      "bindEvents",
    ],
    "cwd": "",
    "dataSaving": false,
    "directives": Object {
      "EventListener": [Function],
      "disabled": [Function],
      "onclick": [Function],
      "oninput": [Function],
      "onload": [Function],
      "onresize": [Function],
      "onscroll": [Function],
      "onselect": [Function],
      "onsubmit": [Function],
    },
    "name": "",
    "onSubmit": [Function],
    "renderAsync": false,
    "renderDependence": Array [],
    "saveButtonText": "",
    "types": Array [],
  },
}
`;
